### ISSUE TITLE
Unexpected NaNs in a jax port of PSIS-LOO

### ISSUE THREAD

**Issue by adamhaber** [google/jax#12952]
### Description

I'm trying to port the Pareto smoothed importance sampling (PSIS) algorithm to JAX, since I need to be able to differentiate through the log-weights; specifically, I am interested in porting the `psislw` function. There are already two existing implementations I'm aware of - Aki Vehtari's original [version](https://github.com/avehtari/PSIS/blob/master/py/psis.py), and an Arviz [version](https://arviz-devs.github.io/arviz/_modules/arviz/stats/stats.html#psislw).

The main difficulty in porting these to JAX was that the original code involves a lot of dynamic array operations -- shapes depend on values. To try and solve this, I wrote a version in which all array sizes are static and "truncations" are replaced with `jnp.nan`s (so that sums and means could be replaced by `jnp.nansum` and `jnp.nanmean`, etc). There were also some control-flow issues which I replaced with `jnp.where` (sometimes at the cost of pre-computing both branches - that may be avoidable, I'm not sure).

The current version of my jax port is [here](https://gist.github.com/adamhaber/0556671340e0daa9e2c6e3fd535cd992).

This worked fine, and I was able to reproduce the results of both implementations on random data. However, when I try to take the `grad` of the version I wrote, I'm getting `nan` gradients; when I ran this with `config.update("jax_debug_nans", True)`, I got this (just the relevant part):

```
~/projects/recast-optimizer-api/optimizer/algorithms/loo.py in psis_single(x, cutoff_ind, cutoffmin, k_min)
    103 
    104     expxcutoff = np.exp(xcutoff)
--> 105     x2 = np.where(x > xcutoff, x, np.nan)
    106     n2 = np.where(x > xcutoff, 1, 0).sum()
    107 
``` 

so seems like the `nan`-padding worked fine for the "forward pass", but doesn't play nicely with the gradient computation? Is that the expected behavior?

Thanks!

### What jax/jaxlib version are you using?

0.3.14

### Which accelerator(s) are you using?

CPU

### Additional system info

macOS, Python 3.9.5

**Comment by shoyer** [comment]
Please take a look at the FAQ entry on "Gradients contain NaN where using `where`":
https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where

**Comment by adamhaber** [comment]
Thanks for the quick response @shoyer !

If I understand this correctly, this means that the idea of "nan-padding" to bypass the dynamic nature of arrays in the `psislw` function isn't going to work, right? Will the same hold for "inf-padding", I guess?

**Comment by hawkinsp** [comment]
I think Stephan's point was not that NaN-padding won't work, but that you may need (surprisingly) two `where` clauses one before and one after.

**Comment by apaszke** [comment]
Yes, this is unfortunately a well known issue with most AD systems. The FAQ section also has links to two issues where it has been explained it detail. The short story is that nan padding is not a bad idea by itself, but you might need to replace those nans with values from the domain of a function before applying it. You might get away with not doing that for some basic stuff like `+`, but anything more complicated (`*`, `sin`, `exp`, ...) will likely break.

I'll close this since it's already been discussed at length. But feel free to open up a new discussion if you need more help!
